File: Dynamic Programming/2684.Maximum Number of Moves in a Grid/Solution.java
class Solution {
    public int maxMoves(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] f = new int[m][n];
        for (int[] r : f) {
            Arrays.fill(r, Integer.MAX_VALUE);
        }
        for (int i = 0; i < m; i++) {
            f[i][0] = grid[i][0];
        }
        int ans = 0;
        for (int j = 1; j < n; j++) {
            for (int i = 0; i < m; i++) {
                int v = f[i][j - 1];
                if (i > 0) {
                    v = Math.min(v, f[i - 1][j - 1]);
                }
                if (i < m - 1) {
                    v = Math.min(v, f[i + 1][j - 1]);
                }
                if (grid[i][j] > v) {
                    f[i][j] =  grid[i][j];
                    ans = j;
                }
            }
        }
        return ans;
    }
}

File: Dynamic Programming/2684.Maximum Number of Moves in a Grid/Solution.py
class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        f = [[inf] * n for _ in range(m)]
        for i in range(m):
            f[i][0] = grid[i][0]
        ans = 0
        for j in range(1, n):
            for i in range(m):
                v = f[i][j - 1]
                if i > 0:
                    v = min(v, f[i - 1][j - 1])
                if i < m - 1:
                    v = min(v, f[i + 1][j - 1])
                if grid[i][j] > v:
                    f[i][j] = grid[i][j]
                    ans = j
        return ans